### ISSUE TITLE
Extend github workers for `MSVC`.

### ISSUE THREAD

**Issue by marcfehling** [dealii/dealii#13596]
It appears that the library has some issues with the `MSVC` compiler as reported in #13593.

We could catch these problems earlier when we extend the github workers accordingly:

- [ ] compile examples
cmake reports `DEAL_II_COMPONENT_EXAMPLES = ON`, but I see no indication of the examples being compiled in the logs. Source files will be copied to the installation folder, but the examples will never be compiled.
- [ ] introduce a parallel build (at least with MPI)

**Comment by marcfehling** [comment]
We could also use this [docker image](https://github.com/mstorsjo/msvc-wine) to run MSVC on Linux via wine. It sounds absurd though :-)

**Comment by marcfehling** [comment]
Related to #12263 and #13593.

**Comment by marcfehling** [comment]
> cmake reports DEAL_II_COMPONENT_EXAMPLES = ON, but I see no indication of the examples being compiled in the logs. Source files will be copied to the installation folder, but the examples will never be compiled.

I found out that we explicitly disable the compilation of examples on Windows systems since #3986.
https://github.com/dealii/dealii/blob/f7f9f468f8d0b1c427c8bf2f938ebbd22ca6414d/cmake/checks/check_02_system_features.cmake#L115-L116

@tamiko Do you recall why this was necessary?

**Comment by tjhei** [comment]
With static linking you would get 100s of MB per binary built, which will also take a very long time. Maybe that's the reason?

**Comment by marcfehling** [comment]
I have removed the `set(DEAL_II_COMPILE_EXAMPLES OFF)` line just to see how it affects our github-runners. I think compiling the examples is done in a feasible time, but the resulting archive size grows dramatically.

We can exclude the examples from the archive to keep its size small:
1. Configure, build and install the library without examples, and then build the examples target separately
2. -or- Erase the examples directory before creating the archive

I would opt for option 2. What do you think?


#### windows-2019

|                                                                                                       | runtime build_library | runtime build_examples |
| ----------------------------------------------------------------------------------------------------- | --------------------- | ---------------------- |
| static library, no examples, [demo](https://github.com/dealii/dealii/actions/runs/9886847557)         | 15m 57s               | 0s                     |
| static library, build examples, [demo](https://github.com/marcfehling/dealii/actions/runs/9878743765) | 27m 5s                | **6m 1s**              |


#### windows-2022

|                                                                                                        | runtime build_library | runtime build_examples | size of archive |
| ------------------------------------------------------------------------------------------------------ | --------------------- | ---------------------- | --------------- |
| static library, no examples, [demo](https://github.com/dealii/dealii/actions/runs/9886847557)          | 14m 44s               | 0s                     | 337MB           |
| static library, build examples, [demo](https://github.com/marcfehling/dealii/actions/runs/9878743765)  | 26m 51s               | **8m 6s**              | **1.07GB**      |

**Comment by marcfehling** [comment]
I was also experimenting with shared libraries on Windows as an alternative to static linking. So I enabled `BUILD_SHARED_LIBS` for Windows. That lead to plenty of unresolved symbols errors of the following kind:
```
step-xx.obj : error LNK2019: unresolved external symbol ...
```

Then I found out that `CMake >=3.4` features a [`WINDOWS_EXPORT_ALL_SYMBOLS`](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) property, which would be helpful in that regard. So I enabled [`CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`](https://cmake.org/cmake/help/latest/variable/CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS.html) as well which triggered a limitation of the MSVC linker.
```
LINK : fatal error LNK1189: library limit of 65535 objects exceeded
```

[Microsoft](https://learn.microsoft.com/en-us/cpp/error-messages/tool-errors/linker-tools-error-lnk1189) suggests to split the library into multiple ones. This is annoying. We could specify a different linker and experiment with that. Maybe we can tackle that at a later stage.

**Comment by bangerth** [comment]
Yes, shared libs on Windows are very solidly stuck in the paleolithic period. I cannot figure out why they have never managed to make the shared lib symbol table larger :-(

`WINDOWS_EXPORT_ALL_SYMBOLS` is the right way forward regardless of whether that leads to a workable solution or not. Want to write a separate patch for that?

**Comment by marcfehling** [comment]
> `WINDOWS_EXPORT_ALL_SYMBOLS` is the right way forward regardless of whether that leads to a workable solution or not. Want to write a separate patch for that?

I just saw that @tamiko already tried that in #14567 without success :-(

**Comment by tjhei** [comment]
> We can exclude the examples from the archive to keep its size small:

I agree with removing them from the archive. if you run setup_tests_examples and ctest instead, are they then part of the archive?

**Comment by marcfehling** [comment]
Building the examples via `DEAL_II_COMPILE_EXAMPLES` works fine. However, building the `setup_tests_examples` target and running `ctest` currently leads to numerous issues, see for example:
https://github.com/marcfehling/dealii/actions/runs/10040414637

I am currently investigating the issue. I uploaded the complete `build` folder as a artifact, which you can find following the above link. https://github.com/dealii/dealii/pull/17278#issuecomment-2233541289 suspected something that source files might be empty after patching.

I can't confirm that yet, as the server terminates the download halfway through. Not sure if artifacts are allowed to be this large (4GB). In hindsight, just uploading the `tests` folder would be enough...

**Comment by tjhei** [comment]
I wouldn't be surprised if @tamiko s Cmake and diff magic doesn't work on windows.

**Comment by marcfehling** [comment]
You can find the test folder from the build directory here. It is still large because of static linking.
https://github.com/marcfehling/dealii/actions/runs/10043129469

The folder `examples/source` constains the patched source files at the top-level, as well as folders for the individual examples.

I indeed noticed that patched source files are empty, but only __some__ of them.

It turns out that if the diff file is empty, the patched file will also be empty. This leads to errors of the following type:
```
  Start 15: examples/step-1.debug

Test command: "C:\Program Files\CMake\bin\cmake.exe" "-DTRGT=examples.step-1.debug.test" "-DTEST=examples/step-1.debug" "-DEXPECT=PASSED" "-DBINARY_DIR=D:/a/dealii/dealii/build/tests/examples" "-P" "D:/a/dealii/dealii/build/share/deal.II/scripts/run_test.cmake"
Working Directory: D:/a/dealii/dealii/build/tests/examples/binary/step-1.debug/serial
Test timeout computed to be: 600
Test examples/step-1.debug: BUILD
===============================   OUTPUT BEGIN  ===============================
examples/step-1.debug: BUILD failed. Output:
MSBuild version 17.10.4+10fbfbf2e for .NET Framework

  1>Checking Build System
  Patching D:/a/dealii/dealii/build/tests/examples/source/step-1.cc
  Building Custom Rule D:/a/dealii/dealii/build/tests/examples/source/CMakeLists.txt
  step-1.cc
MSVCRTD.lib(exe_main.obj) : error LNK2019: unresolved external symbol main referenced in function "int __cdecl invoke_main(void)" (?invoke_main@@YAHXZ) [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-1.debug.vcxproj]
D:\a\dealii\dealii\build\tests\examples\binary\step-1.debug\Debug\step-1.debug.exe : fatal error LNK1120: 1 unresolved externals [D:\a\dealii\dealii\build\tests\examples\binary\examples.step-1.debug.vcxproj]


examples/step-1.debug: ******    BUILD failed    *******

===============================    OUTPUT END   ===============================
```

**Comment by marcfehling** [comment]
Patching happens here if I understood correctly.
https://github.com/dealii/dealii/blob/5eebebc98c91a77141c1aed558957f4fc2109d88/tests/examples/CMakeLists.txt.in#L15-L21

Would it be possible to just copy the source file if the diff file is empty?
